crates/bevy_ecs/src/storage/table.rs

```diff
@@ -82,15 +109,37 @@ impl TableRow {
 
     /// Creates a `TableRow`.
     #[inline]
-    pub const fn new(index: usize) -> Self {
+    pub const fn from_u32(index: u32) -> Self {
+        Self(index)
+    }
+
+    /// Creates a `TableRow` from a [`usize`] index.
+    ///
+    /// # Panics
+    ///
+    /// Will panic if the provided value does not fit within a [`u32`] with `debug_assertions`.
+    #[inline]
+    pub const fn from_usize(index: usize) -> Self {
+        debug_assert!(index as u32 as usize == index);
         Self(index as u32)
     }
 
-    /// Gets the index of the row.
+    /// Gets the index of the row as a [`usize`].
+    ///
+    /// # Panics
+    ///
+    /// Will panic if the inner value does not fit within a [`usize`] with `debug_assertions`.
     #[inline]
-    pub const fn index(self) -> usize {
+    pub const fn as_usize(self) -> usize {
+        debug_assert!(self.0 as usize as u32 == self.0);
```

Comment:
Since we don't support 16 bit platforms (we don't even allow bevy to compile on 16-bit), a `u32` to `usize` cast will always be successful for both 32-bit and 64-bit platforms. We don't really need this assert.